src/raft/raft.go

```diff
@@ -618,20 +793,39 @@ func checkApplyEntries(rf *Raft) {
 	}
 }
 
+func (rf *Raft) committedEqualsApplied() bool {
+	rf.mu.Lock()
+	defer rf.mu.Unlock()
+	return rf.lastApplied == rf.committedIndex
+}
+
 func applyEntries(rf *Raft) {
 	rf.cond.L.Lock()
 	for {
-		for rf.lastApplied == rf.committedIndex {
+		for rf.committedEqualsApplied() {
 			rf.cond.Wait()
 		}
+		rf.mu.Lock()
 		commandIndex := rf.lastApplied + 1
+		// It's possible that rf.me is not in majority.
+		if commandIndex > rf.getLastLogIndex() {
+			rf.mu.Unlock()
+			continue
+		}
+		rf.debug("Server [%d] (start index: [%d], last log index: [%d]) applying log [%d]", rf.me, rf.logs.StartIndex, rf.getLastLogIndex(), commandIndex)
+		if rf.getPartialLogIndex(commandIndex) >= len(rf.logs.Entries) {
+			rf.debug("On server [%d], real index [%d] for command [%d] is greater than max log entry [%d]", rf.me, rf.getPartialLogIndex(commandIndex), commandIndex, len(rf.logs.Entries)-1)
+		}
```

Comment:
Looks like this log is used to debug a specific problem, instead of for general event information. Should probably remove it.